pulpcore/pulpcore/app/viewsets/content.py

```diff
@@ -61,6 +61,22 @@ def create(self, request):
         headers = self.get_success_headers(serializer.data)
         return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
 
+    def destroy(self, request, pk):
+        """
+        Remove Content unit only if it is not associated with a repository version
+        """
+        repository_content_set = RepositoryContent.objects.filter(content__pk=pk,
+                                                                  version_removed=None)
```

Comment:
I think `version_removed` might not have to be `None` and the content still thought of as being associated with a repository; see e.g the [RepositoryVersion.content()](https://github.com/pulp/pulp/blob/3.0-dev/pulpcore/pulpcore/app/models/repository.py#L275,#L279)

Reply:
Ok.  If content may be added, removed, then re-added in another version, do we wish to track all prior version_added/version_removed data?  If no, we can have version_removed changed back to None when re-adding content and this would still be fine.  If yes, we may want this stored in a different data structure or handled differently.

Either way, both the code you linked to earlier regarding relationships:  https://github.com/pulp/pulp/blob/3.0-dev/pulpcore/pulpcore/app/models/repository.py#L275,#L279
and the code in remove_content:  https://github.com/pulp/pulp/blob/c36ab038100ebdc23d23654901dde5f7425fb242/pulpcore/pulpcore/app/models/repository.py#L412-L416
make assumptions that the content is not re-added by filtering out those with version_removed dates without checking how they relate to the current version_added date (the former) and by assuming version_removed to be None (the latter).

Reply:
The short answer is yes—we want to track all the information about content being added/removed/re-added/etc. If a RepositoryContent record exists then the content is in some RepositoryVersion (regardless of whether version_removed is None or not).

We store the changes between RepositoryVersions instead of storing a direct relationship between RepositoryVersions and content. If you're curious about why this is, here's some info:

https://www.redhat.com/archives/pulp-dev/2017-December/msg00023.html

Reply:
Thanks!